File: day05.py
"""Solution to day 5 for Advent of Code"""

from get_input import get_input
from copy import deepcopy
from itertools import takewhile, count
import re


def part1(stack_instructions):
    stacks, instructions = stack_instructions
    stacks = deepcopy(stacks)
    for (cnt, frm, to) in instructions:
        for _ in range(cnt):
            stacks[to].append(stacks[frm].pop())
    return ''.join(stacks[s][-1] for s in sorted(stacks.keys()))


def part2(stack_instructions):
    stacks, instructions = stack_instructions
    stacks = deepcopy(stacks)
    for (cnt, frm, to) in instructions:
        stacks[to].extend(stacks[frm][-cnt:])
        stacks[frm][-cnt:] = []
    return ''.join(stacks[s][-1] for s in sorted(stacks.keys()))


def parse(text):
    crate_lines = []
    lines = iter(text.strip().split("\n"))
    for line in takewhile(lambda line: line != "", lines):
        crate_lines.append(line)
    cols = [int(n) for n in crate_lines[-1].strip().split(' ') if n != '']
    crates = {c: [] for c in cols}
    for line in crate_lines[:-1]:
        for i, key in zip(count(1, 4), cols):
            if line[i] == " ":
                continue
            assert 'A' <= line[i] <= 'Z'
            crates[key].insert(0, line[i])
    instructions = []
    move_re = re.compile(r"^move (\d+) from (\d+) to (\d+)$")
    for line in lines:
        m = move_re.match(line)
        instructions.append(tuple(int(g) for g in m.groups()))
    return (crates, instructions)


if __name__ == "__main__":
    LINES = parse(get_input(day=5, year=2022))
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day05.rs
use common::get_input;
use regex::Regex;

struct Instruction {
    from: usize,
    to: usize,
    boxes: usize,
}

impl Instruction {
    fn from_str(text: &str) -> Instruction {
        let re = Regex::new(r"move (\d+) from (\d+) to (\d+)").unwrap();
        let groups = re.captures(text).unwrap();
        let get_int = |n| groups.get(n).unwrap().as_str().parse::<usize>().unwrap();
        Instruction {
            boxes: get_int(1),
            from: get_int(2),
            to: get_int(3),
        }
    }
}


struct Input {
    boxes: Vec<Vec<char>>,
    instructions: Vec<Instruction>
}

impl Input {
    fn from_str(crates: Vec<&str>, instructions: Vec<&str>) -> Input {
        let crate_names = crates.last().unwrap().trim().split("   ")
            .map(|n| n.trim().parse::<usize>().unwrap()).collect::<Vec<usize>>();
        let mut boxes = vec![vec![]; crate_names.len()];
        for line in crates.iter().take(crates.len()-1) {
            let line = line.chars().enumerate().filter(|(n, _)| (n + 3) % 4 == 0);
            for (n, (_, chr)) in line.enumerate() {
                if chr != ' ' { boxes[n].insert(0, chr) }
            }
        }
        Input {
            boxes,
            instructions: instructions.iter().map(|i| Instruction::from_str(i)).collect()
        }
    }
}

fn part1(input: &Input) -> String {
    let mut crates = input.boxes.clone();
    for instruction in &input.instructions {
        for _ in 0..instruction.boxes {
            let crt = crates[instruction.from-1].pop().unwrap();
            crates[instruction.to-1].push(crt)
        }
    }
    crates.iter().map(|c| c.last().unwrap()).collect::<String>()
}

fn part2(input: &Input) -> String {
    let mut crates = input.boxes.clone();
    for instruction in &input.instructions {
        let len = crates[instruction.from-1].len();
        let mut crts = crates[instruction.from-1].drain((len-instruction.boxes)..len).collect::<Vec<char>>();
        crates[instruction.to-1].append(&mut crts)
    }
    crates.iter().map(|c| c.last().unwrap()).collect::<String>()
}

fn parse(text: String) -> Input {
    let mut lines = text.trim().split("\n").into_iter();
    let crates = lines.by_ref().take_while(|&l| l != "").collect();
    let instructions: Vec<&str> = lines.collect();
    Input::from_str(crates, instructions)
}

fn main() {
    let input = parse(get_input(05, 2022));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
